### DISCUSSION TITLE
After `sendParent`, child ends up incorrectly in final state

### DISCUSSION THREAD

**Discussion by jmakeig** [statelyai/xstate#3134] [Q&A] [answered]
**TL;DR:** After a `sendParent()` action, the sending spawned machine ends up unexpectedly in the final state and thus is no longer interactive. _How do I send something to the parent without triggering a transition in either parent or child?_

## Details

In my app I have a parent collection that lists its children. A user can select a child to edit. As they make changes, the child maintains mutation (clean/dirty) and validation state. When the user is done editing, they can save the child and then either make more edits or close the selected child and choose another one to edit. A classic list-detail UI.

I have implemented this as a parent machine that spawns child machines, one at a time as the user selects them. The guts of the child machine handles the nitty gritty of edits and validation. Once the user is done editing they child, they can persist their changes. After the child has been successfully persisted (invoked service) it sends an event to the parent that it has been updated, so the parent can update its list view to reflect any changes, for example, if the user edited the display name of the detail item. 

All of this seems to work perfectly except that after the child sends the event and the parent’s actions complete, the child somehow ends up in its _final state_. I‘ve kept a careful log of the events in both the parent and child and neither appear to transition out of their respective states immediately before or after the child issues `sendParent`.

When I try to update the child again after saving and notifying the parent, I get the error, `Warning: Event "update" was sent to stopped service "Item". This service has already reached its final state, and will not transition.` in the browser console. Curiously, right before that I also get `Warning: No implementation found for action type 'clear_validation'`. `clear_validation` is an action on the child machine. I can’t tell if these two warnings are related.

## Reproduction

I’ve tried to reproduce this with simpler machines, but can’t seem to tickle the same behavior except running in my app. (I think there’s [also a bug in the visualizer](https://github.com/statelyai/xstate-viz/issues/326) that prevents testing communication between spawned actors.)

### Parent

When the user selects a child to edit, the parent transitions to the `selected` state and spawns a child machine. As the child is being edited, the parent remains in the `selected` state. When it receives an `updated_item` event, it updates the list of children in the context to reflect the persisted child.

https://github.com/jmakeig/svelte-xstate-collection-details/blob/main/src/routes/items/_items.js#L66-L78:
 
```js
// …
  unselected: {},
  selected: {
    on: {
      deselect: {
        target: "unselected",
        actions: ["clearSelection"],
      },
      updated_item: {
        // https://spectrum.chat/statecharts/general/actions-without-transitions~43393e12-6989-4bf1-a5ca-eb53752ca8ae
        internal: true,
        actions: ["update_item"], // Implementation below
      },
    },
  }
// …
```

I’ve specified that this is an internal transition, just for good measure. I don’t want the `selection` state to change, just to update the items in the context.

https://github.com/jmakeig/svelte-xstate-collection-details/blob/main/src/routes/items/_items.js#L125-L132

```js
//…
  update_item: assign({
    items: (context, { item }) =>
      context.items.map((existing) => {
        // console.log(item, existing);
        if (existing.itemid === item.itemid) return item;
        else return existing;
      }),
  })
//…
```

### Child

https://github.com/jmakeig/svelte-xstate-collection-details/blob/main/src/routes/items/_item.js#L211-L228
```js
//…
  saving: {
    invoke: {
      src: "persist",
      id: "persist",
      onDone: [
        {
          target: "#Item.initialized.editing",
          actions: ["log", "store", "notify"], // Implementation below
        },
      ],
      onError: [
        {
          actions: "error",
          target: "#Item.error",
        },
      ],
    },
  }
//…
```

The `notify` action uses `sendParent()` to convey the updated child to the parent machine. 

https://github.com/jmakeig/svelte-xstate-collection-details/blob/main/src/routes/items/_item.js#L278
```js
//…
notify: sendParent((context, event) => ({ type: 'updated_item', item: event.data }))
//…
```

After the `onDone` has successfully completed, the child looks to have successfully transitioned back to its `editing` state. However, at some point immediately afterwards it ends up in its final state. I don’t see any other transitions, though.

Any ideas what’s going on or how I can debug this?

Any help would be greatly appreciated.

**Comment by Andarist** [comment]
I'm pretty sure that part of the problem is here:
https://github.com/jmakeig/svelte-xstate-collection-details/blob/538e1580efe5a92e6cedc584fe37ceb151bb96e4/src/routes/items/_items.js#L121

This will stop the child and it won't be able to receive any new events from now on. Perhaps this is a problem because you are sending something to that child through a regular action and those are "deferred" to the end of the current transition, while this `.stop()` call is **eager** and gets executed immediately. 

If we consider code like this:
```js
actions: [
  send({ type: 'FOO' }, { to: ctx => ctx.childRef }),
  assign({
    childRef: (ctx) => {
      ctx.childRef.stop()
      return null
    }
  })
]
```
then what actually happens is that we first stop the child and then try to send an event to it, even though it looks like it should be the other way around.

Using `.stop()` like this is not correct - but also it's not like you have done something wrong here. This isn't well-documented and we should improve the docs around this.

A proper way to do this would look something like this:
```js
actions: [
  send({ type: "FOO" }, { to: (ctx) => ctx.childRef }),
  actions.stop((ctx) => ctx.childRef),
  assign({
    childRef: null
  })
]
```

Unfortunately, this doesn't quite cut it yet - because how `assign` actions are prioritized in the current version of XState. In the next version, this won't be the case and this would just work but right now you need to opt-into the new behavior with:
```js
const parent = createMachine({
  preserveActionOrder: true,
  // ...
})
```

You can check out a working example here:
https://codesandbox.io/s/flamboyant-fast-hlfnjz?file=/src/index.js

I'm not saying that this will instantly resolve your issue but it might and I would start with this refactor. If this doesn't help then I would really love to see a runnable repro case. I see that you have linked to a public repo here so perhaps you could list all repro steps, from how I should start the project to what I should do in tests or in a running app to reproduce your problem.

**Comment by jmakeig** [comment]
Thanks for the quick follow-up, @Andarist. 

Unfortunately, the transition that contains the action that stops the child machine is never hit in my current code. It was speculative. I’ve removed the `on.deselect` and `clearSelection` action entirely and still see the same behavior. 

I mocked out the database access layer, so you should be able to run it locally with recent version of Node:

```shell
git clone https://github.com/jmakeig/svelte-xstate-collection-details.git
cd svelte-xstate-collection-details
npm install
npm run dev -- --open
```

which should open the Svelte Kit app on port 3000 locally. It will redirect to `/items` and launch the XState inspector in a pop-up window. (The hot reloading in Svelte Kit takes some huffing and puffing to start up, so you might need a hard reload on the items page to actually generate the parent items list UI.)

I’ve also instrumented the parent and child machines to show the live state and context as well as a history of the state changes. 

You can see a live version in a [production build deployed on Vercel](https://svelte-xstate-collection-details.vercel.app/items). 

To reproduce:
1. Select an item from the list.
2. Edit the item. The “Save” and “Cancel” buttons should be enabled because it correctly transitions into the `dirty` state. (It’s actually a `dirty` and `valid` parallel state, but that’s not germane.)
3. Save the edited item. This will persist the item and then notify the parent machine with `sendParent()`.
4. Try to edit the item again. In development mode you’ll get the final state error above in the console. In prod, there’s no console message, but the behavior is the same. I would have expected the child to have transitioned back to the `editing.clean` state and be editable. 

Again, thanks for your help. Any debugging hints would be much appreciated.

**Comment by jmakeig** [comment]
I’m not sure if the missing action warning is related, e.g. https://github.com/statelyai/xstate/pull/2539#discussion_r693939091

**Comment by jmakeig** [comment, marked as answer]
Eureka!

For those keeping score at home, my problem was that the custom [Svelte store](https://svelte.dev/docs#run-time-svelte-store) that I use to wrap my interpreted machines was calling [`service.stop()`](https://github.com/jmakeig/svelte-xstate-collection-details/blob/c7c76e973c142db7dbf9f79a0bbcb14053f0ddef/src/lib/service-store.js#L40) in the store’s unsubscribe function. This was a misunderstanding on my part about how store subscriptions work.

Thanks for pointing me in the right direction, @Andarist, with the reference to a different call to `stop()` in my code.

**Reply by Andarist** [reply]
> This was a misunderstanding on my part about how store subscriptions work.

Could you expand on this? When I took a look at this line of code - it seems completely legit and what I would have expected to be called there. So perhaps the issue is something else? Happy to jump on a call to clear this up if you'd like that.

**Reply by jmakeig** [reply]
The return from the store subscription callback is a function that’s called when the subscriber count goes from 1 to 0. It’s not clear to me how that’s related to the lifecycle of my machine. Even if no one is listening the service should remain valid, no? And yet, that’s exactly what the out-of-the-box [useMachine()](https://github.com/statelyai/xstate/blob/b0d268ebaf1be192495e174a8b6de4332dcf6b48/packages/xstate-svelte/src/useMachine.ts#L75) function does. Something else must be going on with my [store wrapper](https://github.com/jmakeig/svelte-xstate-collection-details/blob/c7c76e973c142db7dbf9f79a0bbcb14053f0ddef/src/lib/service-store.js#L24).

**Reply by Andarist** [reply]
The lifecycle of the machine passed to `useMachine` should be bound to the component that "owns" it. So it seems that you are right - this should be refactored. I'm no Svelte expert but when I take a look at this it feels like `readable` is being misused in the implementation. 

I've prepared a small refactor [here](https://github.com/statelyai/xstate/pull/3172/files). Could you check it out? Perhaps you might see a better way of doing this. Before landing this we'd also have to add some tests for this.

**Reply by jmakeig** [reply]
In a very cursory test, calling `service.unsubscribe()` in my store’s unsubscribe callback seems to work. Well, in as much as not doing anything in the store’s unsubscribe callback worked from my particular use case too. Calling unsubscribe definitely seems more correct than `service.stop()`. Presumably unwinding the subscriptions reduces the potential for memory leaks by holding onto subscriber references. I’m not sure how to test this, though.
